### ISSUE TITLE
SerialChart.handleSeriesRemoved incorrectly loops over array while deleting items

### ISSUE THREAD

**Issue by killroy42** [amcharts/amcharts4#2233]
In the file SerialChart.js inside the function `SerialChart.prototype.handleSeriesRemoved` the code uses a `.each` loop while removing items. So, for example, if there are two items, the loop first reaches the first item, then deletes it from the array. On the next iteration, it tries to reach the second item, which now returns undefined (as it now occupies the first position in the array, resulting in a `Uncaught TypeError: Cannot read property 'dataContext' of undefined` error.

This would easily be resolved by either looping over a shallow copy of the array, or looping in reverse. There seems to be an unnecessary double wrapping of a simple for loop within `OrderedList.each` and `$array.each`

One fix could be to replace this line:
`this.legend.dataItems.each(function (dataItem) {`
with
`[...this.legend.dataItems].forEach(function (dataItem) {`

**Comment by killroy42** [comment]
Additionally, this code:
`$array.each(this.legend.data, function (item) {
	$array.remove(_this.legend.data, item);
});`
seems to be functionally equivalent to:
`this.legend.data.length = 0;`
Not sure why all the extra layers of function calls and loops.

**Comment by zeroin** [comment]
Could you provide a codepen which would reproduce the error you mentioned?

**Comment by killroy42** [comment]
I'm afraid I lack the knowledge as to what combination of calls and used features fills in the `this.legend.dataItems` and `dataItem.dataContext` elements, etc. But you can look at the code directly and confirm for yourself that it cannot loop correctly over an array while removing items in that way.

**Comment by killroy42** [comment]
Here is a version of the relevant code broken down for you. I've determined that `legend.dataItems` is implemented via a `SortedList`, which uses `$array.each` and $array.removeIndex` under the bonnet.
So this replicates the problem:

```
function each(array, fn) {
	var length = array.length;
	for (var i = 0; i < length; ++i) {
		fn(array[i], i);
	}
}
function removeIndex(array, index) {
	array.splice(index, 1);
}

function List() {
	this._values = [];
}
List.prototype.each = function(fn) {
	each(this._values, fn);
};
List.prototype.remove = function(item) {
	var index = this._values.indexOf(item);
	removeIndex(this._values, index);
};

var list = new List();
list._values = [{prop: 1}, {prop: 2}, {prop: 3}];
console.log(list._values);

list.each(function(item) {
	console.log('Values before remove:', list._values);
	console.log('item:', item);
	console.log('item.prop:', item.prop);
	list.remove(item);
	console.log('Values after remove:', list._values);
});
```

**Comment by zeroin** [comment]
Fixed this in 4.9.9 - thank you for suggestion!
